### PR TITLE
7 executionlog

### PR DESCRIPTION
**Pull request by pontuslaestadius**

### CODE REVIEW

File: Application/src/net/Decode.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+     * Removes all the characters provided from the string provided.
+     * @param string the string to purify.
+     * @param characters The main card, "Exodia the Forbidden One",
+     *                   is one of the first five Effect Monsters
+     *                   ever to be released and the first card to
+     *                   have an alternative victory condition.
+     * @return a String without any of the characters provided.
+     */
+    private String removeCharactersFromString(String string, char... characters) {
+        StringBuilder sb = new StringBuilder();
```
Comment (ghost) [inline Application/src/net/Decode.java]: maybe it is just me but it is hard to read the code when i have to keep track of the very concise instance names, please make them more self explanatory e.g. UserName userName   instead of UserName un

File: Application/src/net/Decode.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        write(classes);
+    }
+
+    /**
+     * Adds a message to a Draw object.
+     * @param d draw to add the message too.
+     * @param message content and what to display to send.
+     * @param values includes the nodes it traverses.
+     */
+    private void message(Draw d, String message, String[] values) {
```
Comment (ghost) [inline Application/src/net/Decode.java]: here too please

File: Application/src/Main.java [hunk 3]
```diff
@@ -95,17 +84,22 @@ public void changed(ObservableValue<? extends Tab> ov, Tab t, Tab t1) {
...
         );
 
+        ExecutionLog executionLog = new ExecutionLog();
+
         BorderPane borderpane = new BorderPane(); // Initializes a new BorderPane that holds all Elements.
         HBox menu = new HBox(); // A HBox holds items horizontally like a menu.
         menu.getChildren().add(btn_import); // Adds the buttons to the menu box.
-        menu.getChildren().add(btn2);
+        menu.getChildren().add(button_previous); // Adds next button
+        menu.getChildren().add(button_next); // Adds previous button
```
Comment (timjon) [inline Application/src/Main.java]: These 2 lines are breaks the entire program, please fix it

File: Application/src/Main.java [hunk 4]
```diff
@@ -122,7 +116,15 @@ public void changed(ObservableValue<? extends Tab> ov, Tab t, Tab t1) {
         primaryStage.setScene(main);
         primaryStage.show();
 
-        Draw.temp_generate_diagram(); // TODO replace with actual parsing.
-        Resizer.init(primaryStage); // Starts a listener for handling resizing of the window.
+        // https://docs.oracle.com/javafx/2/threads/jfxpub-threads.htm
+        // https://stackoverflow.com/questions/21083945/how-to-avoid-not-on-fx-application-thread-currentthread-javafx-application-th
+        // https://docs.oracle.com/javase/8/javafx/api/javafx/application/Platform.html#runLater-java.lang.Runnable-
+        Platform.setImplicitExit(true);
+        Platform.runLater(() -> {
```
Comment (timjon) [inline Application/src/Main.java]: What do you use these 2 lines for

File: Application/src/net/Decode.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        String id_string = raw.substring(1, id_index);  // Retrieves the Diagram_ID.
+        int id = Integer.decode(id_string); // String to Integer.
+
+        if (raw.contains("simulation_finished")) { // If the simulation is finished.
+            write("Simulation finished"); // Write Simulation finished in the execution log.
+
+        } else if (raw.contains("print_information")) { // If it's a statement to only print.
+            write(raw); // Writes the raw data TODO replace with statement provided.
+            Net.push("{" + id + ", send_message}");
```
Comment (timjon) [inline Application/src/net/Decode.java]: Doesnt this line make it so that the client asks for the next message to be simulated?

File: Application/src/net/Server_connection.java [hunk 6]
```diff
@@ -18,14 +19,14 @@
     /**
      * Initializes the connection to the backend server
      */
-    public void Init(){
-       OpenConnection();
+    public void init(){
```
Comment (timjon) [inline Application/src/net/Server_connection.java:22]: :(

File: Application/src/visuals/DiagramView.java [hunk 7]
```diff
@@ -46,15 +52,13 @@ static String getInView() {
...
      */
-    public static DiagramView getDiagramViewInView() {
-        for (DiagramView d: diagramViews) {
-            Tab t = d.getTab();
-            String id = t.getId();
-            if (id.equals(getInView())) {
+    public static DiagramView getDiagramViewInView() throws IllegalStateException {
+        if (diagramViews.size() == 0)
+            throw new IllegalStateException("No views exist");
+        for (DiagramView d: diagramViews)
```
Comment (timjon) [inline Application/src/visuals/DiagramView.java]: Please add comments for the if-statement and the for-loop

File: Application/src/visuals/Draw.java [hunk 8]
```diff
@@ -59,6 +59,26 @@ public void addMessage(int fromNode, int toNode, String name){
 
     }
 
+    public boolean removeMessage() {
```
Comment (timjon) [inline Application/src/visuals/Draw.java]: No comment for this method

File: Application/src/visuals/Draw.java [hunk 9]
```diff
@@ -59,6 +59,26 @@ public void addMessage(int fromNode, int toNode, String name){
...
+    public int findClassIndex(String name){
+        for (int i = 0; i < diagramClasses.size(); i++) {
+            if (diagramClasses.get(i).getName().equals(name))
+                return i;
+        }
+        return -1;
+    }
+
     // Always renders with a new specific resolution.
     void resize(double w, double h) {
```
Comment (timjon) [inline Application/src/visuals/Draw.java]: java doc

Comment (timjon) [reply Application/src/visuals/Draw.java]: i know this wasnt added now but still

File: Application/src/visuals/Draw.java [hunk 10]
```diff
@@ -106,18 +126,10 @@ void init() {
     void renderContainer() {
         if (!DiagramView.inView(this)) return;
         GraphicsContext gc = canvas.getGraphicsContext2D();
-        Thread c = new Thread(new Render(gc, diagramClasses));
-        c.start();
-
+        for (Renderable r: diagramClasses)
```
Comment (timjon) [inline Application/src/visuals/Draw.java:147]: Where do you get diagramClasses from?

File: Application/src/visuals/ExecutionLog.java [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    private static ExecutionLog elog;
+    private ListView<String> listView = new ListView<>();
+    private ObservableList<String> data = FXCollections.observableArrayList();
+
+
+    public static ExecutionLog getInstance() {
+        return elog;
+    }
+
+    public ExecutionLog() {
```
Comment (timjon) [inline Application/src/visuals/ExecutionLog.java]: You missed comments for 2 methods here

File: Application/src/visuals/ExecutionLog.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * Displays any information given in to a ListView.
+ * @author Pontus Laestadius
+ * @version 0.6
+ */
+public class ExecutionLog extends ListView {
+    private static ExecutionLog elog;
+    private ListView<String> listView = new ListView<>();
+    private ObservableList<String> data = FXCollections.observableArrayList();
```
Comment (timjon) [inline Application/src/visuals/ExecutionLog.java]: What are these 3 used for?

File: Application/src/visuals/handlers/Animation.java [hunk 13]
```diff
@@ -10,6 +10,22 @@
  */
 public class Animation extends Thread {
     private static Thread singletonAnimationThread;
+    private long timeSinceLastUpdate = 0; // Variable used to pinpoint the update time so it's always the same.
+    private static int framesPerSecond = 6; // Defaults framesPerSecond the application updates in.
```
Comment (timjon) [inline Application/src/visuals/handlers/Animation.java]: Should this be final for now?

File: Application/src/visuals/handlers/Animation.java [hunk 14]
```diff
@@ -26,16 +42,24 @@ public void run() {
...
+            try { // Catches if there is not view.
+                // Retrieves the draw that is in view and updates and redraws it.
+                Draw draw = DiagramView.getDiagramViewInView().getDraw(); // Get the Draw object from the view.
+                draw.update(); // Updates the states of the Renderable objects.
+                draw.redraw(); // Redraws their graphics on the canvas.
+            } catch (IllegalStateException ex) {
+                continue;
+            }
+
+            timeSinceLastUpdate = (System.currentTimeMillis() -t1 > 1000/framesPerSecond ?
```
Comment (timjon) [inline Application/src/visuals/handlers/Animation.java]: I know what this line does but maybe add a comment to make it faster to understand

File: Application/src/visuals/handlers/Resizer.java [hunk 15]
```diff
@@ -29,9 +27,9 @@ public void run() {
      */
     private void time(){
         th = this;
-        while ((timer+=inc) < threashold) { // A 400ms delay in which the timer can be reset
+        while ((timer+=5) < 45) { // A 400ms delay in which the timer can be reset
             try {
-                Thread.sleep(inc); // Sleep for the same time as timer increases
+                Thread.sleep(5); // Sleep for the same time as timer increases
```
Comment (timjon) [inline Application/src/visuals/handlers/Resizer.java]: Whats the reasons for the changes in this file?

File: Application/src/net/Net.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Pontus Laestadius
+ * @version 0.6
+ */
+public class Net implements Runnable {
+    private static Server_connection con = null;
+    private static List<String> queue = new ArrayList<>();
```
Comment (timjon) [inline Application/src/net/Net.java]: Isnt there a list type thats made for queues?

**Review by timjon — CHANGES_REQUESTED** [review]
1. The Net class doesnt seem to handle several messages being sent between each next_message command
2.  Whats the random numbers to the left in the execution log
3. The application prints an IndexOutOfBoundsException on the the third time i press the next button (It doesnt seem to break anything)
4. The classes seems to be printed in the execution log when i launch the application
5. There is nothing telling me that the exectution log is the execution log
6. I cant seem to remove the previous in the execution log after a certain point

File: Application/src/net/Decode.java (hunk 5 above)
Comment (pontuslaestadius) [reply Application/src/net/Decode.java]: If it is a print statement, yes. Do you want to change it?

File: Application/src/visuals/handlers/Animation.java (hunk 13 above)
Comment (pontuslaestadius) [reply Application/src/visuals/handlers/Animation.java]: No, there is a setspeed method for external changes.

File: Application/src/Main.java (hunk 3 above)
Comment (pontuslaestadius) [reply Application/src/Main.java]: Works for me, post a screenshot please.

File: Application/src/net/Decode.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+     * @param string the string to purify.
+     * @param characters The main card, "Exodia the Forbidden One",
+     *                   is one of the first five Effect Monsters
+     *                   ever to be released and the first card to
+     *                   have an alternative victory condition.
+     * @return a String without any of the characters provided.
+     */
+    private String removeCharactersFromString(String string, char... characters) {
+        StringBuilder sb = new StringBuilder();
+        boolean e = false;
```
Comment (Sebastian-Fransson) [inline Application/src/net/Decode.java]: Variable names could be more explanatory in this method.

File: Application/src/Main.java (hunk 3 above)
Comment (timjon) [reply Application/src/Main.java]: Check the comments

File: Application/src/net/Decode.java (hunk 5 above)
Comment (timjon) [reply Application/src/net/Decode.java]: Im wondering since the send_message command will queue the next message to be simulated and that will cause the message queue on the client side to stack up messages

File: Application/src/visuals/handlers/Animation.java (hunk 13 above)
Comment (timjon) [reply Application/src/visuals/handlers/Animation.java]: Oh i must have missed that

File: Application/src/visuals/ExecutionLog.java (hunk 12 above)
Comment (Sebastian-Fransson) [reply Application/src/visuals/ExecutionLog.java]: +1

**Review by Sebastian-Fransson — CHANGES_REQUESTED** [review]
Small naming stuff

File: Application/Application.iml [hunk 18]
```diff
@@ -7,5 +7,6 @@
     </content>
     <orderEntry type="inheritedJdk" />
     <orderEntry type="sourceFolder" forTests="false" />
+    <orderEntry type="library" name="com.google.code.gson:gson:2.8.2" level="project" />
```
Comment (IsabelleTornqvist) [inline Application/Application.iml]: Why is this added here? isn't this part of the parsing?

File: Application/src/Main.java [hunk 19]
```diff
@@ -64,26 +58,19 @@ public void handle(ActionEvent event) {
...
-        Button btn2 = new Button();
-        btn2.setText("Settings");
-        btn2.setOnAction(new EventHandler<ActionEvent>() {
-            @Override
-            public void handle(ActionEvent event) {
-                System.out.println("Settings"); } // Settings button handler.
+        Button button_previous = new Button("Previous");
+        button_previous.setOnAction((ActionEvent event) ->{
+            Net.push("{1, previous_message}");
+            // Only go back if you can remove a message.
```
Comment (IsabelleTornqvist) [inline Application/src/Main.java:64]: What happens if you can't?

File: Application/src/Main.java [hunk 20]
```diff
@@ -95,17 +82,22 @@ public void changed(ObservableValue<? extends Tab> ov, Tab t, Tab t1) {
                         for (DiagramView dv: DiagramView.diagramViews) {
                             dv.redraw();
                         }
+                        DiagramView dv = DiagramView.getDiagramViewInView();
```
Comment (IsabelleTornqvist) [inline Application/src/Main.java]: Optimal variable name? (dv)

Comment (IsabelleTornqvist) [reply Application/src/Main.java]: Suggestion: diagramView or similar?

File: Application/src/Main.java [hunk 21]
```diff
@@ -95,17 +82,22 @@ public void changed(ObservableValue<? extends Tab> ov, Tab t, Tab t1) {
...
                 }
         );
 
+        ExecutionLog executionLog = new ExecutionLog();
+
         BorderPane borderpane = new BorderPane(); // Initializes a new BorderPane that holds all Elements.
         HBox menu = new HBox(); // A HBox holds items horizontally like a menu.
         menu.getChildren().add(btn_import); // Adds the buttons to the menu box.
-        menu.getChildren().add(btn2);
+        // menu.getChildren().add(button_previous); // Adds next button
```
Comment (IsabelleTornqvist) [inline Application/src/Main.java]: If this code is commented out, should it be removed?

File: Application/src/net/Decode.java [hunk 22]
```diff
@@ -0,0 +1,159 @@
...
+import visuals.ExecutionLog;
+
+import static visuals.DiagramView.tabPane;
+
+/**
+ * @author Pontus Laestadius
+ * @version 1.0
+ */
+class Decode {
+    String raw; // Raw string to be decoded.
```
Comment (IsabelleTornqvist) [inline Application/src/net/Decode.java]: Comments go above the code, right?

File: Application/src/visuals/Renderable.java [hunk 23]
```diff
@@ -4,6 +4,7 @@
 
 /**
  * Implementing this assures the class can be rendered through generics.
+ * @version 1.0
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Renderable.java]: Author tag?

File: Application/src/visuals/handlers/Resizer.java [hunk 24]
```diff
@@ -7,20 +7,18 @@
 /**
  * The Resizer runs on a separate thread and there can only exist one at a time.
  * It's in charge of when the canvases are meant call resize after a certain time delay.
- * @version 0.5
+ * @version 1.0
  * @author Pontus Laestadius
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/handlers/Resizer.java]: Author tag goes above version tag

File: Application/src/net/Decode.java [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+            } else { // New Diagram
+                diagramClasses(message);
+            }
+        }
+    }
+
+    /**
+     * Removes all the characters provided from the string provided.
+     * @param string the string to purify.
+     * @param characters The main card, "Exodia the Forbidden One",
```
Comment (IsabelleTornqvist) [inline Application/src/net/Decode.java]: What?

File: Application/src/net/Decode.java [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+     *                   is one of the first five Effect Monsters
+     *                   ever to be released and the first card to
+     *                   have an alternative victory condition.
```
Comment (IsabelleTornqvist) [inline Application/src/net/Decode.java]: Not tru, <insert funny comment on this>

File: Application/src/net/Decode.java (hunk 25 above)
Comment (timjon) [inline Application/src/net/Decode.java]: ~

File: Application/src/net/Decode.java [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+     * @param string the string to display in the execution log.
+     */
+    private void write(String string) {
+        Platform.runLater(() -> {
+            // Gets the instance and writes a new line.
+            ExecutionLog.getInstance().fwd(string);
+        });
+    }
+
+    private String retrieveMessage(String string) {
```
Comment (timjon) [inline Application/src/net/Decode.java]: Java doc

File: Application/src/net/Net.java [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+    /**
+     * Pushes a string to queue which will be processed in the order they are received.
+     * @param string to be sent to the server.
+     */
+    public static void push(String string) {
+        // TODO validate message here.
+        queue.add(string);
+    }
+
+    private void inner_loop() {
```
Comment (timjon) [inline Application/src/net/Net.java]: java doc

File: Application/src/net/Net.java [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+            try { // Sleep for for a longer duration before attempting to reconnect.
+                Thread.sleep(3000);
+            } catch (InterruptedException e) {
+                System.err.println(e);
+            }
+        }
+    }
+
+    // Test example to send a diagram to the server.
+    public static void test() {
```
Comment (timjon) [inline Application/src/net/Net.java]: java doc

File: Application/src/net/Decode.java [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+        int msg_start = string.indexOf("["); // Finds the end of the message content.
+        int msg_end = string.indexOf("]"); // Finds the start of the message content.
```
Comment (IsabelleTornqvist) [inline Application/src/net/Decode.java]: The other way around yeah? start, end etc

File: Application/Application.iml (hunk 18 above)
Comment (ghost) [reply Application/Application.iml]: it is because Pontus imported the library to his branch yesterday, let's use this iml file in develop and delete the rest

File: Application/src/net/Decode.java (hunk 25 above)
Comment (ghost) [reply Application/src/net/Decode.java]: @IsabelleTornqvist ikr

File: Application/src/net/Net.java [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+     * Pushes a string to queue which will be processed in the order they are received.
+     * @param string to be sent to the server.
+     */
+    public static void push(String string) {
+        // TODO validate message here.
+        queue.add(string);
+    }
+
+    private void inner_loop() {
+        Server_connection con = new Server_connection();
```
Comment (IsabelleTornqvist) [inline Application/src/net/Net.java]: Revise variable name. Suggestion: serverConnection, server_connection, connection

**Review by timjon — CHANGES_REQUESTED** [review]

File: Application/src/net/Net.java [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+
+            try { // Sleep for for a longer duration before attempting to reconnect.
+                Thread.sleep(3000);
+            } catch (InterruptedException e) {
+                System.err.println(e);
+            }
+        }
+    }
+
+    // Test example to send a diagram to the server.
```
Comment (IsabelleTornqvist) [inline Application/src/net/Net.java]: Is this bellow necessary?

**Review by IsabelleTornqvist — CHANGES_REQUESTED** [review]

File: Application/src/net/Decode.java (hunk 26 above)
Comment (pontuslaestadius) [reply Application/src/net/Decode.java]: :O how is Exodia not true?

File: Application/src/net/Net.java [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+    public static void test() {
+        Net.push("{1, [\"gateway:g\",\"user:u1\",\"user:u2\",\"user:u3\"], " +
```
Comment (IsabelleTornqvist) [inline Application/src/net/Net.java]: Is this necessary?

File: Application/src/visuals/DiagramView.java [hunk 34]
```diff
@@ -33,7 +39,7 @@ public Draw getDraw() {
      */
     static boolean inView(Draw draw) {
         DiagramView dv = getDiagramViewInView();
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/DiagramView.java]: Optimal variable name?

File: Application/src/net/Decode.java (hunk 26 above)
Comment (IsabelleTornqvist) [reply Application/src/net/Decode.java]: it aint

File: Application/src/visuals/Draw.java [hunk 35]
```diff
@@ -3,15 +3,15 @@
...
-import visuals.handlers.Render;
 
 import java.util.ArrayList;
 
 import static visuals.DiagramView.tabPane;
 
 /**
- * @version 0.7
+ * @version 1.0
  * @author Pontus Laestadius, Sebastian Fransson
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Draw.java]: Author tag goes above version tag

File: Application/src/visuals/Draw.java [hunk 36]
```diff
@@ -59,7 +59,34 @@ public void addMessage(int fromNode, int toNode, String name){
...
+    /**
+     * @param name class name to match.
+     * @return the index the DiagramClass is located at.
+     */
+    public int findClassIndex(String name){
+        for (int i = 0; i < diagramClasses.size(); i++) {
+            if (diagramClasses.get(i).getName().equals(name))
+                return i;
+        }
+        return -1;
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Draw.java]: why -1?

File: Application/src/visuals/Draw.java [hunk 37]
```diff
@@ -59,7 +59,34 @@ public void addMessage(int fromNode, int toNode, String name){
...
+    }
+
+    /**
+     * Always renders with a new specific resolution.
+     * @param w the new width of the canvas.
+     * @param h he new height of the canvas.
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Draw.java]: the, not he

File: Application/src/visuals/Draw.java [hunk 38]
```diff
@@ -106,18 +133,10 @@ void init() {
     void renderContainer() {
         if (!DiagramView.inView(this)) return;
         GraphicsContext gc = canvas.getGraphicsContext2D();
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Draw.java]: variable name. Suggestions: graphicsContext

File: Application/src/visuals/Draw.java [hunk 39]
```diff
@@ -181,12 +203,17 @@ void example_diagram(int nr) {
     }
 
     public static void temp_generate_diagram() { // Init's a draw object that handles graphical elements
+        Net.test();
+        //old_generate_diagram();
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/Draw.java]: remove this comment?

File: Application/src/visuals/handlers/Animation.java [hunk 40]
```diff
@@ -26,16 +42,25 @@ public void run() {
...
                 System.err.println(e.toString());
             }
-            // Retrieves the draw that is in view and updates and redraws it.
-            DiagramView dv = DiagramView.getDiagramViewInView();
-            if (dv == null) continue; // If a View is not established. Skip.
-            Draw draw = dv.getDraw(); // Get the Draw object from the view.
-            draw.update(); // Updates the states of the Renderable objects.
-            draw.redraw(); // Redraws their graphics on the canvas.
+
+            long t1 = System.currentTimeMillis();
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/handlers/Animation.java]: t1??

File: Application/src/visuals/handlers/Resizer.java [hunk 41]
```diff
@@ -7,20 +7,18 @@
...
     private static int timer = 0;
-    private int inc = 5;
-    private int threashold = 45;
 
     /**
      * Runs when the thread starts.
      */
     public void run() {
-        if (th == null) // Only allow one instance of the program to run at a time
+        if (th == null) // Only allow one instance to run at a time
```
Comment (IsabelleTornqvist) [inline Application/src/visuals/handlers/Resizer.java]: th?

**Review by IsabelleTornqvist — CHANGES_REQUESTED** [review]

**Review by Sebastian-Fransson — APPROVED** [review]

File: Application/src/visuals/Draw.java (hunk 37 above)
Comment (pontuslaestadius) [reply Application/src/visuals/Draw.java]: Did I just assume it's gender?

File: Application/src/visuals/Draw.java (hunk 38 above)
Comment (pontuslaestadius) [reply Application/src/visuals/Draw.java]: That wasn't changed in this pull request? So I don't think I'll change it.

**Review by timjon — APPROVED** [review]

File: Application/src/visuals/Draw.java (hunk 38 above)
Comment (IsabelleTornqvist) [reply Application/src/visuals/Draw.java]: ops ye sorry

**Review by IsabelleTornqvist — APPROVED** [review]
